fix(scheduler): add failed execution with logs when scheduled triggers fail during evaluation

MTarek165 opened the pull request:
closes #13307
## Description
- Now with scheduled trigger when there is issue at trigger evaluation( unsupported expressions, flow inputs resolving issues ...)
it fails silently without creating failed execution, you will now only from logs that there is a problem
## Root Cause
- They all pass through evaluateScheduleTrigger() method to evaluate trigger and hence the rendering and resolving inputs processes happen, so when any exception from there happen it is propagated to this method and caught in the catch block creating logs but ignoring sending failed execution
## Solution
- Whenever an exception caught from this method create a failed execution and attach logs to them 
<img width="800" height="287" alt="fail1" src="https://github.com/user-attachments/assets/fe917234-5d59-4218-b2b0-757e3bdc5a2d" />
<img width="800" height="362" alt="fail2" src="https://github.com/user-attachments/assets/d30e9e19-eb7f-46b4-a389-848a7fea442a" />

MTarek165 committed:
fix(core): add failed execution with logs when scheduled triggers fail during evaluation

MTarek165 committed:
fix(core): return backfill check at handleFailedEvaluatedTrigger()

loicmathieu commented on scheduler/src/main/java/io/kestra/scheduler/AbstractScheduler.java:
```diff
@@ -712,17 +712,9 @@ private void handle() {
+                                Optional<SchedulerExecutionWithTrigger> schedulerExecutionWithTrigger = evaluateScheduleTrigger(f, scheduleContext);
                                 if (schedulerExecutionWithTrigger.isPresent()) {
                                     this.handleEvaluateSchedulingTriggerResult(schedule, schedulerExecutionWithTrigger.get(), f.getConditionContext(), scheduleContext);
-                                } else {
-                                    // compute next date and save the trigger to avoid evaluating it each second
-                                    Trigger trigger = Trigger.fromEvaluateFailed(
-                                        f.getTriggerContext(),
-                                        schedule.nextEvaluationDate(f.getConditionContext(), Optional.of(f.getTriggerContext()))
-                                    );
-                                    trigger = trigger.checkBackfill();
-                                    this.triggerState.save(trigger, scheduleContext, "/kestra/services/scheduler/handle/save/on-eval-true/schedule");
```
Why did you remove this part?

MTarek165 replied:
- Because when exception is caught at evaluateScheduleTrigger() it returns Optional.Empty() to handle() entering this section, so instead of calling this logic there I include it at the added function and call from catch block at evaluateScheduleTrigger() before returning the empty object.
- Please take a look on the logic at the added function

MTarek165 replied:
If there is an edge case at moving this code at catch block, I can revert back this change and prepare and emit the excution at catch block in evaluateScheduleTrigger() directly

MTarek165 replied:
Anyway, forget about that, I have reverted it back as I found more better way
- I changed it to create failed execution and emit it directly when evaluateScheduleTrigger() catch the exceptions
 this is more cleaner and no changes on old code 
- Added test for it and now CI is green.

loicmathieu reviewed (changes requested):
It also miss a test

MTarek165 committed:
feat(tests): add more test coverage for trigger evaluation failure
- This covers failures propagated to evaluateScheduleTrigger() in AbstractScheduler class related to ScheduleTrigger such as (Invalid Expressions, Inputs resolving Issues...)

MTarek165 committed:
chore(tests): add small notes to tests

MTarek165 commented:
> It also miss a test

I have added a test now

MTarek165 committed:
chore(scheduler): revert back changes at handle() and create failed execution with emitting it directly from the catch block

MTarek165 committed:
Merge branch 'develop' into fix/add-failed-execution-when-scheduled-trigger-evaluation-fails

MTarek165 committed:
Merge branch 'develop' into fix/add-failed-execution-when-scheduled-trigger-evaluation-fails

loicmathieu commented on scheduler/src/test/java/io/kestra/scheduler/SchedulerScheduleTest.java:
```diff
@@ -489,9 +489,9 @@ void stopAfterSchedule() throws Exception {
+// covers trigger evaluation failures from failed conditions at scheduled triggers
     @Test
-    void failedEvaluationTest() {
+    void failedEvaluation1() {
```
Please name the test more accurately like `failedEvaluationFromFailedCondition()`.

loicmathieu commented on scheduler/src/test/java/io/kestra/scheduler/SchedulerScheduleTest.java:
```diff
@@ -527,6 +527,62 @@ void failedEvaluationTest() {
+    }
+    // covers trigger evaluation failures propagated to evaluateScheduleTrigger() such as invalid expressions, input resolving issues at scheduled triggers...etc
+    @Test
+    void failedEvaluation2() {
```
Please name the test more accuratly, like `failedExecutionFromInvalidExpression()`.

MTarek165 committed:
chore(tests): update namings of evaluation tests

MTarek165 commented:
@loicmathieu test names updated

loicmathieu reviewed (approved):
LGTM, thanks
